Epic 6798 : List Domain Designer Conversion

labkey-bpatel opened the pull request:
- Configure list module for React page.
- Addition of an Experimental feature to toggle on/off the new React page.
- Addition of GetDomainDetailsAction (variant of existing GetDomainAction) to return domain kind specific properties; and related refactoring.
- Deprecate GetDomainAction & update to client api docs.
- Bind of DomainKind specific options/properties to a bean for CreateDomain & SaveDomainAction (generifying of DomainKind and related refactoring).
- Notable List specific changes:  Add/override updateDomain(), modify createDomain(), create non-gwt bean, and other minor changes necessary for the new List Designer.

labkey-bpatel committed:
List Designer - initial commit:
- Configure list module for React page.
- Add Experimental feature to toggle on/off the new React page.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_list_designer_6798

labkey-bpatel committed:
Update devDependencies, include package-lock.json.

labkey-bpatel committed:
ListDesigner.tsx entryPoint setup and initial wrapper for shared component.

labkey-bpatel committed:
Added comment.

labkey-bpatel committed:
Add constructor to ListDesigner. Update EditListDefinitionAction - update nav trail and move props code.

labkey-bpatel committed:
Add API action to get ListDefinition.

labkey-bpatel committed:
Include domain when getting ListDefinition (in order to also get list fields in addition to List property).

labkey-bpatel committed:
Update build.gradle to include dependency on npmInstall command finishing at the platform repository level.

labkey-bpatel committed:
Ignore lists' node_modules and auto-generated files.

labkey-bpatel committed:
Improved failure messages.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Rename ListDesigner to Designer, plus related refactoring.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Added new GetDomainKindAction (variant of existing GetDomainAction) which also returns domain kind specific properties; and related refactoring.

RosalineP committed:
Establishes List Properties Panel. DomainId in API call currently hardwired.

labkey-bpatel committed:
Use designer instead of listDesigner.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Establishes componentDidMount. Includes AdvancedSettings modal (not fully wired)

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_list_designer_6798

labkey-bpatel committed:
Rename GetDomainKindAction to GetDomainDetailsAction. Added new GetListPropertiesAction to be able to get domainId.

labkey-bpatel committed:
Updated package-lock.json

RosalineP committed:
Contains initial work on domainForm + ListDesignerPanels section. Uses Binal's GetListPropertiesAction and GetDomainDetails

RosalineP committed:
Adds List Fields. Other work in progress

RosalineP committed:
(Forgot to bump package.json @labkey.components number)

RosalineP committed:
Updates labkey-ui-components version

labkey-bpatel committed:
Return default empty list when listId is not provided or is null.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for CreateDomain:
- Generified DomainKind.
- Added new method getTypeClass() in DomainKind.
- Update ListDomainKind to implement new method with appropriate bean (ListDomainKindProperties).
- Update rest of DomainKind implementations to use JSONObject (instead of Map<String, Object>) so that it works with generified DomainKind (using Map<String, Object> did not work as desired with generics so resorted to using JSONObject).
(In the future when we convert other DomainKinds, use the approach taken for Lists - see ListDomainKind.getTypeClass() & createDomain()).

labkey-bpatel committed:
Remove unused class pushed by mistake.

labkey-bpatel committed:
Comment & Javadoc for getTypeClass().

labkey-bpatel committed:
Update comment.

labkey-bpatel committed:
Update comment, again.

labkey-bpatel committed:
Minor updates - return default integer enum value instead of text.

labkey-bpatel committed:
Minor updates - match the field type to field type in db.

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Supports load and create list scenarios

labkey-bpatel committed:
Allow binding of DomainKind specific options/properties to a bean for SaveDomainAction & and override updateDomain() for List:
- Change to DomainKind.updateDomain() signature to include nullable DomainKind specific properties/options & refactor callers.
- Update SaveDomain js api to include optional DomainKind specific properties/options.

labkey-bpatel committed:
Add evaluationDependsOn()

RosalineP committed:
Do a permissions check. Update components version

labkey-bpatel committed:
Update client api docs to deprecate LABKEY.Domain.get. Include its replacement - LABKEY.Domain.getDomainDetails.

labkey-bpatel committed:
Update to client api docs.

labkey-bpatel committed:
Added further validation. Improved exception throwing.

labkey-bpatel committed:
Remove check on JSONObject.class

RosalineP committed:
Bump ui-components to newly published version

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

RosalineP committed:
Merge branch 'develop' into fb_list_designer_6798

RosalineP committed:
Bump ui-components.

labkey-bpatel committed:
Use ArrayList to avoid ClassCastException error. (arguments.get("idCols") is an ArrayList and casting to Object[] was throwing ClassCastException)

labkey-bpatel committed:
Update ListDomainKindProperties not to extend Entity; and related refactoring.

RosalineP committed:
Bumping ui-components version

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

labkey-bpatel committed:
Get list via domainId and not name (since changing a name of the list doesn't update the name in exp.domainDescriptor, so using GWTDomain.getName() returns an old name).

cnathe committed:
Item 6639: minor updates to list React setup files and add a README for the module

labkey-jeckels commented on assay/package.json:
```diff
@@ -29,7 +29,7 @@
   },
   "dependencies": {
-    "@labkey/components": "0.23.1"
+    "@labkey/components": "0.25.0-fb-list-designer-6798.2"
```
Don't forget to update before merging

labkey-jeckels commented on api/src/org/labkey/api/exp/property/DomainKind.java:
```diff
@@ -97,22 +105,33 @@
     /**
      * Create a Domain appropriate for this DomainKind.
```
This isn't the right JavaDoc for this method, looks like a copy/paste

labkey-bpatel replied:
Now updated

labkey-jeckels commented on api/src/org/labkey/api/exp/property/DomainKind.java:
```diff
@@ -97,22 +105,33 @@
+     * @param user User
+     * @return Return DomainKindProperties.
+     */
+    abstract public DomainKindProperties getDomainKindProperties(GWTDomain domain, Container container, User user);
```
Should this return DomainKindProperties or T? Do we need the DomainKindProperties interface at all now?

labkey-jeckels replied:
and should this method be @Nullable?

labkey-bpatel replied:
It was my initial implementation, but after adding generics, <DomainKindProperties> is now removed.

labkey-jeckels commented on api/webapp/clientapi/core/Domain.js:
```diff
@@ -301,6 +373,8 @@ LABKEY.Domain.create({
          *       If not supplied, the current container path will be used.
          * @param {boolean} config.includeWarnings set this to true, if server side warnings are desired along with the errors
          *       If this is set to true and there are warnings detected, they will prevent the save from completing successfully
+         * @param {boolean} config.options Domain Kind specific properties.
```
This should be {Object}, not {boolean}

labkey-bpatel replied:
Good eye, now updated.

labkey-bpatel replied:
Also, have added an additional Note and have listed read-only properties for IntList and VarList. 
Let me know what you think! Thanks!

labkey-jeckels replied:
Works for me.

labkey-jeckels commented on api/webapp/clientapi/core/Domain.js:
```diff
@@ -413,7 +488,21 @@ LABKEY.Domain.create({
 * @description An array of objects that each describe a domain field.  Each object has the following properties:
+    *       (Note: Not all properties below are expected to have values or will have values. See {@Link LABKEY.Domain} 'Create' to see minimally required properties when creating a Domain.)
     *      <ul>
+    *          <li><b>defaultDisplayValue:</b> Default value to display. (string)</li>
```
Let's put these newly doc'd properties at the bottom of the list - they're less likely to be relevant to the developer than the others.

labkey-bpatel replied:
Makes sense, now updated

labkey-jeckels commented on internal/src/org/labkey/api/issues/AbstractIssuesListDefDomainKind.java:
```diff
@@ -69,7 +70,7 @@
  * Created by davebradlee on 8/3/16.
  */
-public abstract class AbstractIssuesListDefDomainKind extends AbstractDomainKind
+public abstract class AbstractIssuesListDefDomainKind<T> extends AbstractDomainKind<JSONObject>
```
This doesn't need its own <T> - all subclasses use JSONObject. This should clean up some of the casting in other spots in the code.

labkey-bpatel replied:
Yup, now removed.

labkey-jeckels commented on list/src/org/labkey/list/controllers/ListController.java:
```diff
@@ -246,45 +250,73 @@ public ActionURL getRedirectURL(ListDefinitionForm listDefinitionForm)
 
+    @Marshal(Marshaller.Jackson)
+    @RequiresPermission(ReadPermission.class)
+    public class GetListPropertiesAction extends ReadOnlyApiAction<ListDefinitionForm>
```
Do we still need this or can we just use the new general-purpose domain API?

labkey-bpatel replied:
I think this was added for the new list designer to get list properties from a listId, it is currently being used in react code.

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,266 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+            ListDefinition listDefinition = ListService.get().getList(domain);
+            TableInfo table = listDefinition.getTable(user);
+
+            if (null == domain || null == table)
```
If domain is null, I assume the call to ListService.getList() will blow up

labkey-bpatel replied:
Good point, now updated

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,266 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+            {
+                int id = key.getPropertyId();
+                GWTPropertyDescriptor newKey = findField(id, update.getFields());
+                if (null != newKey && !key.getName().equalsIgnoreCase(newKey.getName()))
```
If newKey == null, shouldn't that be a problem too?

labkey-bpatel replied:
Perhaps I am missing something here -- this code is in updateDomain(), a null newKey is what's expected since we don't want key field name to be changed during update (and its a problem only if newKey is not null and not same as existing key).

labkey-bpatel replied:
As per our discussion, this is check is now in place. Thanks for clarifying!

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,266 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+                }
+                catch (RuntimeSQLException e)
+                {
+                    if (changedName && e.isConstraintException())
```
Better to check for this up front instead of catching an exception

labkey-bpatel replied:
its now updated by Cory as part of a bug fix.

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,266 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+    @NotNull
+    private ValidationException checkLegalNameConflicts(GWTDomain dd)
+    {
+        ValidationException errors = new ValidationException();
```
Do we need this here. Shouldn't the centralized Domain saving code handle this?

labkey-bpatel replied:
Now removed, since duplicate name check is already happening in DomainUtil.validateProperties().

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,266 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+    {
+        ListDomainKindProperties updatedListProps = new ListDomainKindProperties(existingListProps);
+
+        if (null != newListProps.getName() && !newListProps.getName().equals(updatedListProps.getName()))
```
Why bother checking if it's equal? It's just as easy to stomp over the old value with an identical one.

labkey-bpatel replied:
True, good point, now updated.

labkey-bpatel replied:
Additionally, have removed read-only properties such as keyName, keyType, and lastIndexed.

labkey-jeckels commented on study/src/org/labkey/study/model/AbstractSpecimenDomainKind.java:
```diff
@@ -53,7 +54,7 @@
 import java.util.Set;
 
-public abstract class AbstractSpecimenDomainKind extends AbstractDomainKind
+public abstract class AbstractSpecimenDomainKind<T> extends AbstractDomainKind<JSONObject>
```
Same as the issue domain kind - no need for this class to take a T

labkey-bpatel replied:
I think it was giving me a compile time error given my changes, hence I had included it. But now removed after adding a convenience subclass 'BaseAbstractDomainKind' of AbstractDomainKind.

labkey-jeckels commented on study/src/org/labkey/study/query/studydesign/AbstractStudyDesignDomainKind.java:
```diff
@@ -48,7 +49,7 @@
  * Created by klum on 12/10/13.
  */
-public abstract class AbstractStudyDesignDomainKind extends AbstractDomainKind
+public abstract class AbstractStudyDesignDomainKind extends AbstractDomainKind<JSONObject>
```
What do you think about introducing a convenience subclass of AbstractDomainKind that is tied to JSONObject, so that and using that here and other places so we don't need to implement getTypeClass() in so many places? Not sure it's worth it, but could be more convenient.

labkey-bpatel replied:
Good idea, thanks for suggesting. Meet 'BaseAbstractDomainKind', subclass of AbstractDomainKind. And it was worth it.

labkey-jeckels reviewed (changes requested):
I didn't review the .ts or .tsx files, but here are the suggestions for the Java changes

cnathe committed:
Item 6639: List designer updates to use functions from new shared component package

cnathe committed:
Item 6639: fix issue with save not getting listId based on the container

cnathe committed:
Use labkey primary button styling for domain designers in LKS pages

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

cnathe committed:
Item 6639: list designer support for dirty state check on navigate
- also add query to get listId for newly saved list for navigating to list grid

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Item 6639: update to latest @labkey/components alpha package version

cnathe commented on api/webapp/clientapi/core/Domain.js:
```diff
@@ -252,6 +263,65 @@ LABKEY.Domain.create({
+        alert('An error occurred retrieving data: ' + error);
+     }
+
+    LABKEY.Domain.get(successHandler, errorHandler, 'study', 'StudyProperties');
```
Looks like there as still a few places in this doc info that reference "get" instead of "getDomainDetails", here and up above for the success and failure callback info.

labkey-bpatel replied:
Good catch, now updated!

cnathe committed:
Item 6801: For new list creation, apply options/properties to the newly created list definition in ListDomainKind.createDomain

cnathe committed:
Item 6639: Update to latest @labkey/components alpha package version

cnathe committed:
Item 6801: Fix for duplicate name check on list designer create and update

labkey-bpatel committed:
Code Review Changes:
- Corrections to javadocs and client api docs.
- Remove DomainKindProperties interface, and related refactoring.
- Add convenience subclass of AbstractDomainKind named BaseAbstractDomainKind, and related refactoring.

cnathe committed:
Item 6801: update list grid "view design" link text for new designer (i.e. just use "design" for link text)

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Update to latest @labkey/components alpha package version after merge from master

cnathe committed:
Update to latest @labkey/components alpha package version

labkey-bpatel committed:
Code Review change: Removed name conflicts check since its already being checked in centralized domain code. Update list properties except listId, domainId, keyName and keyType.

labkey-bpatel committed:
Code review changes:
- Handle cases when existing key field is null or is not provided in the updated domainDesign.
- Remove lastIndexed property from being updated.
- Update client api doc listing read-only props for List.
- Set empty string as default values for entireListTitleTemplate & eachItemTitleTemplate properties.

labkey-bpatel committed:
Slight rewording.

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Update to latest @labkey/components alpha package version

labkey-jeckels commented on internal/src/org/labkey/api/exp/property/BaseAbstractDomainKind.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.json.JSONObject;
+
+public abstract class BaseAbstractDomainKind<T> extends AbstractDomainKind<JSONObject>
```
This shouldn't take a <T> - T is never used

labkey-bpatel replied:
Now updated!
(Actually that's what I had initially, but I was getting some error around that, I must not have updated something correctly since it builds fine now)

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -441,37 +441,47 @@ public ValidationException updateDomain(GWTDomain<? extends GWTPropertyDescripto
+                if (null == newKey)
                 {
-                    exception.addError(new PropertyValidationError("Cannot change key field name", key.getName()));
+                    throw new IllegalArgumentException("Key field not provided, expecting key field '" + key.getName() + "'");
```
Why are these IllegalArgumentExceptions instead of adding to the errors object? And if it makes sense to not use errors, please use ApiUsageException instead to avoid mothership logging.

labkey-bpatel replied:
Now added to the error obj for updateDomain().
CreateDomain() was using IllegalArugmentException as well, but now those are changed to use ApiUsageException.

RosalineP committed:
Bumping component version

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,245 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+            updatedListProps.setTitleColumn(newListProps.getTitleColumn());
+
+        if (null != newListProps.getDescription())
+            updatedListProps.setDescription(newListProps.getDescription());
```
Does this mean that you can't completely clear out a description once it's been set?

labkey-bpatel replied:
AH! Excellent point! (You alluded to this earlier as well, but didn't quiet take that further).
Null checks are now removed for all props except name.

labkey-jeckels commented on list/src/org/labkey/list/model/ListDomainKind.java:
```diff
@@ -410,6 +432,245 @@ public Domain createDomain(GWTDomain domain, Map<String, Object> arguments, Cont
+                GWTPropertyDescriptor newKey = findField(id, update.getFields());
+                if (null == newKey)
+                {
+                    throw new IllegalArgumentException("Key field not provided, expecting key field '" + key.getName() + "'");
```
Why are these IllegalArgumentExceptions instead of adding to the errors object? And if it makes sense to not use errors, please use ApiUsageException instead to avoid mothership logging.

labkey-bpatel replied:
Done!

labkey-jeckels reviewed (changes requested):
Close, but a few more suggestions

labkey-bpatel committed:
Change to isEmpty() check instead of null check.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

labkey-bpatel committed:
Code review change : Remove unused generic type, and related refactor.

labkey-bpatel committed:
Code Review changes:
- Use ApiUsageException instead of IllegalArgumentException for CreateDomain.
- Add to ValidationException error obj instead of using IllegalArgumentException.
- Remove null check when updating props to allow for change from a non-null to null val (except for name).

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6798: one more update to Domain.js docs

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Update to latest @labkey/components alpha package version

RosalineP committed:
Bumping component version

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	experiment/package-lock.json
#	experiment/package.json

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6639: handle list creation case with error during file import

RosalineP committed:
Bumping components version

RosalineP committed:
Merge remote-tracking branch 'origin/fb_list_designer_6798' into fb_list_designer_6798

# Conflicts:
#	list/package.json

RosalineP committed:
Bumping components version

RosalineP committed:
Bumping components version

cnathe committed:
Item 6798: automated test updates to prep tests that use ListHelper.createList()

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Issue 39808: Fix for domain field validation check to trim before checking that a name is not blank

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6798: update to latest @labkey/components alpha package version

cnathe committed:
Merge branch 'develop' into fb_list_designer_6798

cnathe committed:
Update to @labkey/components release version 0.31.0
